modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractBulkByScrollRequest.java

```diff
@@ -0,0 +1,328 @@
+import static org.elasticsearch.common.unit.TimeValue.timeValueMinutes;
+import static org.elasticsearch.search.sort.SortBuilders.fieldSort;
+
+public abstract class AbstractBulkByScrollRequest<Self extends AbstractBulkByScrollRequest<Self>>
+        extends ActionRequest<Self> {
+    public static final int SIZE_ALL_MATCHES = -1;
+    private static final TimeValue DEFAULT_SCROLL_TIMEOUT = timeValueMinutes(5);
+    private static final int DEFAULT_SCROLL_SIZE = 100;
+
+    /**
+     * Default search source.
+     */
+    private static final BytesReference DEFAULT_SOURCE = new SearchSourceBuilder().version(true)
+            .size(AbstractBulkByScrollRequest.DEFAULT_SCROLL_SIZE).sort(fieldSort("_doc")).buildAsBytes();
+
+    /**
+     * The search to be executed.
+     */
+    private SearchRequest searchRequest;
+
+    /**
+     * Maximum number of processed documents. Defaults to -1 meaning process all
+     * documents.
+     */
+    private int size = SIZE_ALL_MATCHES;
+
+    /**
+     * Should version conflicts cause aborts? Defaults to true.
+     */
+    private boolean abortOnVersionConflict = true;
+
+    /**
+     * Call refresh on the indexes we've written to after the request ends?
+     */
+    private boolean refresh = false;
+
+    /**
+     * Timeout to wait for the shards on to be available for each bulk request?
+     */
+    private TimeValue timeout = ReplicationRequest.DEFAULT_TIMEOUT;
+
+    /**
+     * Consistency level for write requests.
+     */
+    private WriteConsistencyLevel consistency = WriteConsistencyLevel.DEFAULT;
+
+    /**
+     * Initial delay after a rejection before retrying a bulk request. With the default maxRetries the total backoff for retrying rejections
+     * is about one minute per bulk request. Once the entire bulk request is successful the retry counter resets.
+     */
+    private TimeValue retryBackoffInitialTime = timeValueMillis(500);
+
+    /**
+     * Total number of retries attempted for rejections. There is no way to ask for unlimited retries.
+     */
+    private int maxRetries = 11;
+
+    public AbstractBulkByScrollRequest() {
+    }
+
+    public AbstractBulkByScrollRequest(SearchRequest source) {
+        this.searchRequest = source;
+        source.scroll(DEFAULT_SCROLL_TIMEOUT);
+    }
+
+    /**
+     * `this` cast to Self. Used for building fluent methods without cast
+     * warnings.
+     */
+    protected abstract Self self();
+
+    /**
+     * Applies the defaults to the request that cannot be applied during construction. This is super inefficient because it must deserialize
+     * and reserialize the request's source but this is the only way to do it in 2.x.
+     */
+    void applyDefaults() {
+        if (searchRequest.source() == null) {
+            searchRequest.source(DEFAULT_SOURCE);
+        }
+        try {
+            Map<String, Object> newSource = XContentHelper.convertToMap(DEFAULT_SOURCE, true).v2();
```

Comment:
OK! This is one of the places where reindex has to expand the source of the search request so it can set it up. In this case we layer the defaults "under" the search source as sent by the user. This method is called right before we start processing the request on the coordinating node.